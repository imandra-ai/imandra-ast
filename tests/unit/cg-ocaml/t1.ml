module Fmt = CCFormat
module A = Imandra_ast

let decls_json = CCIO.File.read_exn "./t1.json"

let decls =
  match A.Decl.list_of_of_yojson @@ Yojson.Safe.from_string decls_json with
  | Ok d -> d
  | Error e ->
    Fmt.eprintf "cannot decode decls: %s" e;
    assert false

let code = Imandra_ast_cg_ocaml.codegen decls
let () = Fmt.printf "ocaml code:@.```@.%s@.```@." code
