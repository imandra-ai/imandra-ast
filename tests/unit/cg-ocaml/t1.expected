ocaml code:
```
(* generated from imandra-ast *)

open Imandra_prelude;;

type complex = {
  re: Z.t;
  im: Z.t;
}

type point = 
  | Polar of Z.t * Z.t
  | Cartesian of complex


let c_plus = (fun c1 ->
 (fun c2 ->
 {re=( c1.re +  c2.re); im=( c1.im +  c2.im)}))

let is_cartesian = (fun _x ->
 (match _x with
| Polar(_,_) -> false
| Cartesian(_) -> true
))


```
